modules/multisignatures.js

```diff
@@ -80,158 +80,221 @@ function Multisignatures(cb, scope) {
+	);
+	library.bus.message('signature', signature, true);
+	return setImmediate(cb);
+};
 
-		// Find public key
-		let verify = false;
-
-		try {
-			for (
-				let i = 0;
-				i < multisignatureTransaction.asset.multisignature.keysgroup.length &&
-				!verify;
-				i++
-			) {
-				const key = multisignatureTransaction.asset.multisignature.keysgroup[
-					i
-				].substring(1);
-				verify = library.logic.transaction.verifySignature(
-					multisignatureTransaction,
-					key,
-					transaction.signature
-				);
+/**
+ * Process signature for transactions that comes from already estabilished multisignature account
+ *
+ * @private
+ * @param {Object} signature - Signature data
+ * @param {string} [signature.publicKey] - Public key of account that created the signature (optional)
+ * @param {string} signature.transactionId - Id of transaction that signature was created for
+ * @param {string} signature.signature - Actual signature
+ * @param {Object} transaction - Corresponding transaction grabbed from transaction pool
+ * @param {function} cb - Callback function
+ * @returns {setImmediateCallback} cb, err
+ */
+__private.processSignatureFromMultisignatureAccount = (
+	signature,
+	transaction,
+	cb
+) => {
+	// Get sender account of correscponding transaction
+	modules.accounts.getAccount(
+		{ address: transaction.senderId },
+		(err, sender) => {
+			if (err || !sender) {
+				const message = 'Unable to process signature, account not found';
+				library.logger.error(message, { signature, transaction, error: err });
+				return setImmediate(cb, message);
 			}
-		} catch (e) {
-			library.logger.error(e.stack);
-			return setImmediate(cb, 'Failed to verify signature');
-		}
 
-		if (!verify) {
-			return setImmediate(cb, 'Failed to verify signature');
-		}
+			// Assign members of multisignature account from transaction
+			const members = sender.multisignatures;
 
-		return done(cb);
-	}
-	modules.accounts.getAccount(
-		{
-			address: multisignatureTransaction.senderId,
-		},
-		(err, account) => {
-			if (err) {
-				return setImmediate(cb, 'Multisignature account not found');
+			// Check if signature is valid
+			if (!__private.isValidSignature(signature, members, transaction)) {
+				return setImmediate(
+					cb,
+					'Unable to process signature, verification failed'
+				);
 			}
 
-			let verify = false;
-			const multisignatures = account.multisignatures;
+			// Add signature to transaction
+			transaction.signatures.push(signature.signature);
+			// Check if transaction is ready to be processed
+			transaction.ready = Multisignature.prototype.ready(transaction, sender);
 
-			if (multisignatureTransaction.requesterPublicKey) {
-				multisignatures.push(multisignatureTransaction.senderPublicKey);
-			}
+			// Emit events
+			library.network.io.sockets.emit(
+				'multisignatures/signature/change',
+				transaction
+			);
+			library.bus.message('signature', signature, true);
+			return setImmediate(cb);
+		}
+	);
+};
```

Comment:
`__private.processSignatureFromMultisignatureAccount` and `__private.processSignatureForMultisignatureAccountCreation` have a lot of similar code, can we refactor those two to share the same logic.